### PR TITLE
Fix potential hang in right join

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #6528 #6692

Problem Summary:
#6576 introduce a potential bug that for a right join, if probe thread meet error, some probe thread maybe hang forever.
 
### What is changed and how it works?
1. fix potential hang bug in right join
2. refine the sync logical in join for both build and probe. 
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp [hunk 1]
```diff
@@ -120,7 +120,7 @@ void CreatingSetsBlockInputStream::createAll()
             for (auto & elem : subqueries_for_sets)
             {
                 if (elem.second.join)
-                    elem.second.join->setBuildTableState(Join::BuildTableState::WAITING);
+                    elem.second.join->setInitActiveBuildConcurrency();
```
Comment (SeaRise) [inline dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp:123]: Can we move to `PhysicalJoin`?

Comment (windtalker) [reply dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp:123]: Currently, it must be set after compile, see the comments in https://github.com/pingcap/tiflash/pull/6690/files#diff-e6f85ec76cd474ca862de277609c1dfbc11aa70692a3acf3f4e85fefb649cb65R473-R474

**Review by SeaRise — APPROVED** [review]
lgtm

File: dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp [hunk 2]
```diff
@@ -238,13 +238,6 @@ void CreatingSetsBlockInputStream::createOne(SubqueryForSet & subquery)
             }
         }
 
-
-        if (subquery.join)
-        {
-            FAIL_POINT_TRIGGER_EXCEPTION(FailPoints::exception_mpp_hash_build);
```
Comment (gengliqi) [inline dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp:244]: `extern const char exception_mpp_hash_build[];` in this file can also be deleted.

Comment (windtalker) [reply dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp:244]: Done.

File: dbms/src/Interpreters/Join.cpp [hunk 3]
```diff
@@ -1992,16 +1999,51 @@ void Join::checkTypesOfKeys(const Block & block_left, const Block & block_right)
...
+}
+
+void Join::waitUntilAllBuildFinished() const
+{
+    std::unique_lock lk(build_probe_mutex);
+
+    build_cv.wait(lk, [&]() { return meet_error || active_build_concurrency == 0; });
+    if (meet_error) /// throw this exception once failed to build the hash table
+        throw Exception("Build failed before join probe!");
+}
```
[the remark this answers is not in the document]
Comment (windtalker) [reply dbms/src/Interpreters/Join.cpp]: Done

**Review by yibin87 — APPROVED** [review]
LGTM

**Review by gengliqi — APPROVED** [review]
LGTM

**Pull request merged**